### PR TITLE
Add KV cache operation observability

### PR DESCRIPTION
**Pull request by shipiyouniao**
## Operation Accounting

Related to #375. Add exporter-neutral counters for manager operations and bounded error metadata, exposed through the existing vLLM and SGLang snapshot interfaces. Metrics backends remain outside the allocator.

Pool occupancy alone cannot distinguish successful allocation, recoverable capacity rejection, and a broken mapping transaction. A consumer attaching after an error also needs cumulative event counts, not a history entry created only when someone requests a snapshot.

## Execution Path

The manager records requests before allocation, outcomes after allocation, and completed progress during free. The allocation decision and native mapping protocol are unchanged. State-consistency exceptions from #418 still propagate; they are not turned into ordinary capacity rejection.

The implementation is in [`KVCacheManager`](https://github.com/shipiyouniao/kvcached/blob/740c74dd0aa3b4af6fe3f33cc0b72ee06354416d/kvcached/kv_cache_manager.py#L451):

```python
indices = self._alloc_impl(need_size, _skip_wait=_skip_wait)
if indices is None:
    self._increment_operation_counter("allocation_failures_total")
    self._increment_operation_counter("capacity_exhausted_total")
else:
    self._increment_operation_counter("allocation_successes_total")
    self._increment_operation_counter("allocated_blocks_total", len(indices))
```

`manager_page_allocations_total` and `manager_page_releases_total` count manager-page handoffs, including reserved-page reuse. They do not claim to count CUDA physical map/unmap operations: native preallocation, trimming and resizing have a different accounting boundary. Caller-free block counts retain completed progress if a later page release fails. Internal allocation rollback does not inflate those counts. Rejected deferred resize does not count as completed resize.

Before:

```mermaid
sequenceDiagram
    participant E as Engine
    participant M as KV manager
    participant A as Native allocator
    participant C as Metrics consumer
    E->>M: Allocate / free
    M->>A: Existing page operation
    A-->>M: Result or exception
    M-->>E: Existing result
    C->>M: Pool snapshot
    M-->>C: Current occupancy
```

After:

```mermaid
sequenceDiagram
    participant E as Engine
    participant M as KV manager
    participant A as Native allocator
    participant C as Metrics consumer
    E->>M: Allocate / free
    M->>M: Record request
    M->>A: Unchanged page operation
    A-->>M: Result or exception
    M->>M: Record outcome and completed progress
    M-->>E: Preserve result / exception
    C->>M: Operation snapshot
    M-->>C: Cumulative counters and bounded last-error metadata
```

## Concurrency and Cost

Counters use a per-manager lock, not a process-wide history lock. Reading a snapshot copies counters under that same lock. The allocation hot path still incurs counter updates even if no consumer is scraping, so read-only snapshots alone are not evidence of negligible cost. Validation below separates that cost from periodic snapshot collection.

The capability API added by #474 now advertises operation counters and lists the actual cumulative counter names. The list is derived from the snapshot schema, excludes last-error metadata, and leaves the upstream backend and integration capability fields intact.

## Validation

Current head: `740c74dd0aa3b4af6fe3f33cc0b72ee06354416d`, rebased on `main` at `7af4a2245e5cf6ffe4f22bf712a552ac1f8de72b` after #474 merged. The full CI matrix and native GPU fault tests below were rerun on this exact head.

| Check | Result |
| --- | --- |
| C++ CPU suite and full pre-commit | Passed |
| Python 3.9-3.13 CPU suites | 311 passed per version |
| Python 3.9-3.13 MyPy | Passed |
| Native CUDA/VMM regression suite on T4 | 88 passed |
| Actual native allocation-failure injection | One hit; counted once; retry and CUDA readback succeeded |
| Unknown peer result injected at the broadcast callback | State-consistency error propagated and allocator failed closed, without counting ordinary capacity exhaustion |
| Reserved-page reuse and later trim | Two manager handoffs counted; no claim of two physical mappings |
| Post-rebase two-engine vLLM serving regression, no sampling and 10 Hz sampling | 64/64 requests, 8192/8192 output tokens, normal chat and health checks passed; 1241 samples, zero collection errors |

### Allocation Microbenchmark

The performance measurements in this and the following section were collected on the pre-rebase candidate `5a4b3ae99f51365a85d5cc0ae906c6e970509859` against `944ed53399f989afc1317734e689c406cde2b097`. The rebase adds capability discovery integration and tests; `kv_cache_manager.py`, native allocator code and vLLM patches are unchanged from the measured candidate. These are retained measurements, not a new A/B run on the current head.

Real VMM-backed managers on one T4, preallocation and reserved pages disabled. An anchor block keeps the current page mapped, so each timed iteration allocates and frees one other block without paying for a new CUDA mapping. There are 50,000 iterations per round, one discarded warmup and five measured rounds; runs are ordered baseline/candidate/candidate/baseline. The table reports the median of ten measured round averages per configuration. Every run ends with CUDA readback verification.

| Pools | Snapshot rate | Baseline alloc/free pair | This PR |
| --- | --- | --- | --- |
| 1 | None | 21.469 us | 26.729 us |
| 1 | 1 Hz | 21.704 us | 26.610 us |
| 1 | 10 Hz | 21.846 us | 26.725 us |
| 8 | None | 21.592 us | 26.422 us |
| 8 | 1 Hz | 21.799 us | 27.081 us |
| 8 | 10 Hz | 22.156 us | 26.485 us |

The counters add about 5 us per alloc/free pair in this tight loop, roughly 22-25% without sampling. This is a measurable hot-path cost, not a claim of free observability. Periodic sampling does not materially amplify it in these tests. It also does not follow that model serving loses the same percentage: the serving comparison below measures that separately.

### Serving Comparison

One T4 runs two independent TP1 vLLM 0.28.0 V1 engines, Qwen2.5-0.5B-Instruct FP16, Triton attention, CUDA graph capture size 64, synchronous scheduling and prefix caching. Both revisions use 2 MiB pages, preallocation off, zero reserved pages, a 4096-token prefix cap, `KVCACHED_GPU_UTILIZATION=0.9`, and the same explicit 10 GiB logical KV capacity per engine. Engine limits are 8192 model tokens, 2048 batched tokens and 64 sequences. MPS is not enabled.

Each engine uses three warmups, fixed random-dataset seeds, temperature 0 and `ignore_eos`:

- Ordinary load: I256/O128, N32/C4, 4 requests/s.
- Memory pressure: I4096/O128, N64/C32, unlimited arrival rate. A separate fixed CUDA allocation leaves about 3.1 GiB device free before the workload on both revisions.

Runs are baseline/candidate/candidate/baseline, with 768/768 successful benchmark requests and all expected output tokens produced. Health and normal chat probes pass before and after each run. Combined throughput below is total output tokens divided by the longer of the two engines' timed benchmark intervals, averaged over the two runs of each revision.

| Workload | Baseline combined tok/s | This PR combined tok/s | Change | Baseline / PR mean TPOT |
| --- | --- | --- | --- | --- |
| Ordinary | 537.04 | 534.71 | -0.43% | 13.760 / 13.814 ms |
| Memory pressure | 80.883 | 80.508 | -0.46% | 364.643 / 367.866 ms |

These runs show a small end-to-end difference, not the roughly 25% difference in the isolated tight loop. They do not establish zero overhead or statistical equivalence. Sampling runs are reported below separately.

Two further runs sampled live pool and operation snapshots at 1 Hz and 10 Hz. These are single-run sampling checks, not another repeated A/B experiment:

| This PR sampling rate | Ordinary combined tok/s | Pressure combined tok/s | Samples / errors across both engines |
| --- | --- | --- | --- |
| None, two-run mean | 534.71 | 80.508 | Not sampled |
| 1 Hz | 537.43 | 80.850 | 500 / 0 |
| 10 Hz | 534.00 | 80.764 | 4974 / 0 |

Including those checks, all 1152 benchmark requests produced the full 147456 output tokens. No CUDA error, OOM, illegal memory access or EngineCore failure was found in the serving logs. Services were stopped by the test harness after their recovery probes; the T4 returned to idle.

The T4 tests cover single-GPU accounting, readback, error propagation and sampling. They do not establish multi-GPU transaction safety or large-model TP/MPS performance.

### CODE REVIEW

File: kvcached/kv_cache_manager.py [hunk 1]
```diff
@@ -230,6 +266,19 @@ def _wait_post_init(self):
...
+        if counters is None:
+            counters = {}
+            self._operation_counters = counters
+        counters[name] = counters.get(name, 0) + value
+
+    def _record_operation_error(self, code: str, counter_name: str) -> None:
+        self._increment_operation_counter("operation_errors_total")
+        self._increment_operation_counter(counter_name)
+        self._last_error_code = code
+        self._last_error_timestamp_ns = time.time_ns()
```
[the remark this answers is not in the document]
Comment (shipiyouniao) [reply kvcached/kv_cache_manager.py]: Fixed in e6a7c540. Operation counters and last-error fields now use a dedicated `RLock`, including snapshot reads, so the background post-init path is safe even when the allocator uses `NoOpLock` for synchronous scheduling. Added a four-thread counter regression test. Focused tests pass (8 passed), and the full Linux CI matrix passes.

File: kvcached/observability.py [hunk 2]
```diff
@@ -0,0 +1,432 @@
...
+    mapped_bytes = int(manager.get_mapped_memory_size("bytes"))
+    virtual_per_layer_bytes = _int_attr(manager, "mem_size") or 0
+    num_layers = _int_attr(manager, "num_layers") or 0
+    num_kv_buffers = _int_attr(manager, "num_kv_buffers") or 0
+    block_size_bytes = _int_attr(manager, "block_mem_size") or 0
+    bytes_per_block = block_size_bytes * num_layers * num_kv_buffers
+    available_blocks = int(manager.available_size())
+    allocated_blocks = int(manager._get_num_alloced_blocks())
+    reserved_blocks = len(getattr(manager, "reserved_blocks", []))
+
```
[the remark this answers is not in the document]
Comment (shipiyouniao) [reply kvcached/observability.py:263]: Fixed in e6a7c540. Reserved blocks are now reported as a disjoint category and subtracted from both available and allocated values, with non-negative saturation. Snapshot tests cover the updated block and byte accounting, and the full Linux CI matrix passes.

File: kvcached/kv_cache_manager.py [hunk 3]
```diff
@@ -461,10 +557,30 @@ def _pick_avail_page(self, remaining_need: int) -> InternalPage:
...
+                    "free_failed",
+                    "free_errors_total",
+                )
+            raise
+
+        if had_inconsistency:
+            self._increment_operation_counter("free_failures_total")
+        else:
+            self._increment_operation_counter("free_successes_total")
+        self._increment_operation_counter("freed_blocks_total", freed_blocks)
```
Comment (jeff3071) [inline kvcached/kv_cache_manager.py:579]: `_rollback_partial_alloc` calls `free()` and increments `freed_blocks_total`, while `allocated_blocks_total` is not incremented because the allocation returns None. 

This can make freed blocks exceed allocated blocks. 
Could internal rollback free these blocks without incrementing `free_requests_total` and `freed_blocks_total`?

Comment (shipiyouniao) [reply kvcached/kv_cache_manager.py:579]: Good catch. The rollback now calls the internal `_free()` path, so it still releases any physical page obtained by the failed transaction but does not increment the public free request/success/block counters. I also added a regression test for a partial allocation miss that verifies `freed_blocks_total` remains zero while `physical_page_frees_total` records the actual page release.

**Comment by shipiyouniao** [comment]
Rebased on #474 and connected the capability flag/name list to the actual operation-counter schema. Allocation behavior and the native mapping protocol are unchanged.

I also filled the performance-evidence gap. The tight VMM-backed alloc/free loop has a real cost: about 5 us per pair. The matched two-engine T4 serving runs were -0.43% at ordinary load and -0.46% under memory pressure; 1 Hz and 10 Hz snapshot checks had no collection errors. That supports a small serving difference in this setup, not zero overhead or a multi-GPU performance guarantee.

The description now separates the measured pre-rebase SHA from the current SHA. Full CI, native GPU fault/accounting tests and a bounded vLLM serving/sampling regression were rerun after the rebase. No need to infer performance from unit tests alone.

**Comment by shipiyouniao** [comment]
Rebased onto current main, including #456 and #480. The physical-capacity cache invalidations are preserved; a regression test also checks cache invalidation when page initialization fails after a successful allocation.

On `8fb5d2c`, full pre-commit, C++ CPU tests, all five Python CPU/MyPy jobs, and 88 native T4 tests passed. The live allocation-failure, unknown-peer-result and reserved-page-reuse checks passed too. The description separates these new checks from the older performance measurements, which have not been rerun against the new base.

@RixinLiu, could you take another look when you have time? #421 is rebased on this commit and still needs this one to land first.
